src/Symfony/Component/PropertyAccess/PropertyAccessor.php

```diff
@@ -457,13 +458,79 @@ private function writeProperty(&$object, $property, $singular, $value)
 
             $object->$property = $value;
         } elseif (self::ACCESS_TYPE_MAGIC === $access[self::ACCESS_TYPE]) {
-            $object->{$access[self::ACCESS_NAME]}($value);
+            $this->callMethod($object, $access[self::ACCESS_NAME], $value);
         } else {
             throw new NoSuchPropertyException($access[self::ACCESS_NAME]);
         }
     }
 
     /**
+     * Throws a {@see UnexpectedTypeException} as in PHP 7 when using PHP 5.
+     *
+     * @param object $object
+     * @param string $method
+     * @param mixed  $value
+     *
+     * @throws UnexpectedTypeException
+     * @throws \Exception
+     */
+    private function callMethod($object, $method, $value) {
+        if (PHP_MAJOR_VERSION >= 7) {
+            try {
+                $object->{$method}($value);
+            } catch (\TypeError $e) {
+                throw $this->createUnexpectedTypeException($object, $method, $value);
```

Comment:
should we check the error message as done for the E_RECOVERABLE_ERROR below to be sure that we deal really with an exception thrown from $method rather than from some other deeply nested call?